Shared/UserViewModel.swift

```diff
@@ -0,0 +1,92 @@
+//
+//  UserViewModel.swift
+//  Fun Gen
+//
+//  Created by jules on 4/18/22.
+//
+
+import Foundation
+import FirebaseAuth
+import FirebaseFirestore
+import FirebaseFirestoreSwift
+
+class UserViewModel: ObservableObject {
+    @Published var user: User?
+    
+    private let auth = Auth.auth()
+    private let database = Firestore.firestore()
+    var uuid: String? {
+        auth.currentUser?.uid
+    }
+    var userIsAuthenticated: Bool {
+        auth.currentUser != nil
+    }
+    var userIsAuthenticatedAndSynced: Bool {
+        user != nil && userIsAuthenticated
+    }
+    
+    // Firebase Auth Functions
+    func signIn(email: String, password: String) {
+        auth.signIn(withEmail: email, password: password) { [weak self] result, error in
+            guard result != nil, error == nil else {
+                return
+            }
+            DispatchQueue.main.async {
+                self?.sync()
+            }
+        }
+    }
+    
+    func signUp(email: String, firstName: String, lastName: String, password: String) {
+        auth.createUser(withEmail: email, password: password) { [weak self] result, error in
+            guard result != nil, error == nil else {
+                return
+            }
+            DispatchQueue.main.async {
+                self?.add(User(firstName: firstName, lastName: lastName))
+                self?.sync()
+            }
+        }
+    }
+    
+    func signOut() {
+        do {
+            try auth.signOut()
+            user = nil
+        } catch {
+            print("Error signing out user: \(error)")
+        }
+    }
+    
+    // Firestore Functions for User Data
+    // Note: try using snapshot listener instead of calling sync function
```

Comment:
Xcode has a special annotation for this:

```suggestion
    // FIXME: try using snapshot listener instead of calling sync function
```

This will introduce a warning, but I think it's a good idea to have a constant reminder for us that something needs to be fixed

Reply:
Fixed